dartagnan/src/main/java/com/dat3m/dartagnan/Dartagnan.java

```diff
@@ -228,30 +230,44 @@ public static String generateResultSummary(VerificationTask task, ProverEnvironm
             if (hasViolations) {
                 printWarningIfThreadStartFailed(p, encCtx, prover);
                 if (props.contains(PROGRAM_SPEC) && FALSE.equals(model.evaluate(PROGRAM_SPEC.getSMTVariable(encCtx)))) {
-                    summary.append("Program specification violation found").append("\n");
-                    summary.append("User assertion: ").append(p.getSpecification().toStringWithType()).append("\n");
+                    summary.append("===== Program specification violation found =====\n");
+                    for(Event e : p.getEvents(Local.class)) {
+                        if(e.is(Tag.ASSERTION) && TRUE.equals(model.evaluate(encCtx.execution(e)))) {
+                            summary
+                                .append("\t").append(e.getGlobalId())
+                                .append(":\t(").append(e.getSourceCodeFile()).append("#").append(e.getCLine())
+                                .append(")\n");
+                        }
+                    }
+                    summary.append("=================================================\n");
                 }
                 if (props.contains(LIVENESS) && FALSE.equals(model.evaluate(LIVENESS.getSMTVariable(encCtx)))) {
-                    summary.append("Liveness violation found").append("\n");
+                    summary.append("============ Liveness violation found ============\n");
+                    for(Event e : p.getEvents(Label.class)) {
+                        if(e.is(Tag.SPINLOOP) && TRUE.equals(model.evaluate(encCtx.execution(e)))) {
+                            summary
+                                .append("\t").append(e.getGlobalId())
+                                .append(":\t(").append(e.getSourceCodeFile()).append("#").append(e.getCLine())
+                                .append(")\n");
+                        }
+                    }
+                    summary.append("=================================================\n");
```

Comment:
This won't work for two reasons:
- For dynamic spinloops, the label is not tagged.
- Even if the label is tagged, just executing it is not a violation because it's just the header of a potentially spinning loop.

You need to look for jumps that are tagged by `SPINLOOP` and then you need to check if those got executed AND if their condition was also true (for `CondJump` we set the `exec` variable to true no matter if the condition holds or not).
The metadata of those jumps might not be accurate, cause some of them get generated by the loop cutting pass.
I think you either need to somehow find the loop-header and use its metadata or you might check the loop cutting pass and add attach the correct metadata whenever it inserts a new event.

Reply:
The fact that the `SPINLOOP` is executed and its condition is true means that it is stuck as defined in `PropertyEncoder.isStuck`, right?

We only need to propagate the meta data from the label to `assumeSideEffect` in `DynamicPureLoopCutting` if I understood it correctly. This should make it work both for statically and dynamically detected loops.

Reply:
Technically, execution with a true condition does only mean the loop did not terminate in a single iteration, but it does not mean it is stuck. To be stuck, it also needs to read co-maximal stores.
However, if you know that liveness is violated, then it must be stuck for otherwise we would not have reported a violation to begin with. So it is indeed enough to do that simple check.

Adding the correct metadata to `assumeSideEffect` should do the trick. But you might want to only copy over the `cLine`. If you copy all metadata, you would also copy over `oId` and `cId`, which might be misleading.